### ISSUE TITLE
Accurate Scoring based on color input plane hack

### ISSUE THREAD

**Issue by wpstmxhs** [gcp/leela-zero#2059]
Yesterday I completed a program which computes an accurate score from board, even with ~0.01 points accuracy.

I thought that we can score accurate points from a board with changing komi via color input plane hack.

It works with PhoenixGo weights because PhoenixGo's komi seems correct in terms of real score points.

Scoring algorithm is quite simple:

1) First, check for situations with komi=-64.0 and komi=64.0.
If one player wins too much (64+ points), stop to search and print that players wins 64+ points.

2) And then, explore an optimal komi in binary search manner.
We can assume that an optimal komi is a komi with 50.00% winning rate.

And this is an example:

In Ke Jie vs AlphaGo Game 3:
![kejie](https://user-images.githubusercontent.com/28377280/49269344-dda39080-f4a7-11e8-8124-128d6c166b32.png)

In this situation, How many points is Ke Jie losing?
Well, I think that human players can't estimate it because value of center is ambiguous.
However, PhoenixGo can do it.

![kejie_score](https://user-images.githubusercontent.com/28377280/49269406-1d6a7800-f4a8-11e8-9355-f618bc95cafd.png)
It says that Black is winning by 24.13 points without komi.

If someone is interested in it, I can modify it with English and upload its program and codes.

**Comment by l1t1** [comment]
interesting

**Comment by alreadydone** [comment]
This is the same idea as in my #1599 and #1772. If you set `--nonslack --max-wr 0.5 --min-wr 0.5 --wr-margin 0.0`, you'll see two values "komi" and "opp_komi" in console output (and Lizzie will display them). However I don't think the komi would be very accurate, especially at large komi values; after all the neural net hasn't been trained under such komi. I don't think PhoenixGo will play 6.5 komi games very well just by modifying the color plane. In fact, you could get very different komi after black's moves than after white's moves (for many 15b nets at least, forgot about PhoenixGo). My branches can also take search into account to give a more accurate "komi" and doesn't just rely on the winrates at the root node. I also use binary search, but didn't set any upper and lower bounds for komi, and sometimes you can observe 500+ komi IIRC.

**Comment by wpstmxhs** [comment]
@alreadydone 
Yes, that's why I restricted the calculation komi range within -64.0~64.0 range.
PhoenixGo is likely to misunderstand the position with komi larger than say, 100.0.
And in fact, my approach is slightly different from yours.
In your dynamic komi version, the program adjusts komi from inspecting pure network evaluations with different komis.
But In mine, the program lets leela zero think with certain playouts(e.g. 4800 playouts) to evaluate winning rate with a specific komi.

**Comment by wpstmxhs** [comment]
@alreadydone
Yes, that's why I restricted the calculation komi range within -64.0~64.0 range.
PhoenixGo is likely to misunderstand the position with komi larger than say, 100.0.

And for LeelaMaster 15b nets, yes. the network understands concept of komi quite differently from human's scoring concepts. That's why I didn't use 15b nets even if it's lighter than PhoenixGo.

**Comment by alreadydone** [comment]
Oh so you do one search after changing komi each time. I instead collect positions (--adj-positions) from a single search and sample some (--adj-pct, default 4%) of the positions and compute average winrate under differen komi from them. Your approach would produce more accurate results while mine is well suited when you want to spend less resource to determine a reasonable komi.

**Comment by wpstmxhs** [comment]
@alreadydone 
Yeah, you're right.
I added to `think_playouts()` method which thinks with certain playouts, in UCTSearch class.
and for each komi, cleared NNCache and searched again and again.

**Comment by wpstmxhs** [comment]
I changed this program to work in English if the system is not in Korean, and fixed issues with non-ASCII outputs of Korean characters on Sabaki.

Try this.

Compiled binary (with VS2017): [ScoreBot_v1.1_exe.zip](https://github.com/gcp/leela-zero/files/2634655/ScoreBot_v1.1_exe.zip)
Entire source code: 
[ScoreBot_v1.1_src.zip](https://github.com/gcp/leela-zero/files/2634666/ScoreBot_v1.1_src.zip)

To do scoring on certain board, you can type 'phoenixgo-score' command manually or issue an analysis command if you're using GoGui.

**Comment by palusijf** [comment]
Can't compile under macOS.

Scanning dependencies of target objs
[  3%] Building CXX object CMakeFiles/objs.dir/src/GTP.cpp.o
warning: unknown warning option '-Wno-maybe-uninitialized'; did you mean
      '-Wno-uninitialized'? [-Wunknown-warning-option]
/Users/jfmac/Downloads/ScoreBot/src/GTP.cpp:47:10: fatal error: 'Windows.h' file
      not found
#include <Windows.h>
         ^~~~~~~~~~~
1 warning and 1 error generated.
make[2]: *** [CMakeFiles/objs.dir/src/GTP.cpp.o] Error 1
make[1]: *** [CMakeFiles/objs.dir/all] Error 2
make: *** [all] Error 2

**Comment by wpstmxhs** [comment]
@palusijf 
You can comment out the line `#include <Windows.h>`
Sorry for inconvenience.

**Comment by palusijf** [comment]
Thanks

**Comment by wonderingabout** [comment]
@wpstmxhs i assume you use lz-phoenixgo right ?

do you have projects to try this in the phoenixgo software provided by tencent ?
https://github.com/Tencent/PhoenixGo

**Comment by afalturki** [comment]
Thanks. This would work much better when a strong 19x19 Sai network is available I would assume.

**Comment by wpstmxhs** [comment]
@palusijf I'm sorry for late reply.
That error is because of Korean characters.
Try to replace GTP.cpp file with new file included in below file.
[GTP.zip](https://github.com/gcp/leela-zero/files/2640630/GTP.zip)

**Comment by wpstmxhs** [comment]
@wonderingabout Yes, you're right.
Currently I don't have any plan to use the official PhoenixGo software for new thing to try.
But, If I have a such plan, I will take a look into it.
Thanks.

**Comment by wonderingabout** [comment]
@wpstmxhs 

thanks

the dynamic komi works nicely with @alreadydone 's work on leela zero,
so since phoenixgo is a zero bot too, i assume the same idea can be exported to it, that would be very interesting

please keep me updated if you do @wpstmxhs
